File: bridge/workers/worker.R
# SymBridge R worker — an R runtime conducted by Sym.
# JSON lines over stdin/stdout. Talks only to Sym.
suppressMessages(library(jsonlite))

con_in <- file("stdin", open = "r", blocking = TRUE)

respond <- function(x) {
  cat(toJSON(x, auto_unbox = TRUE, null = "null", digits = NA), "\n", sep = "")
  flush(stdout())
}
ok <- function(id, value = NULL, exports = NULL) {
  r <- list(id = id, ok = TRUE, value = value)
  if (!is.null(exports)) r$exports <- exports
  respond(r)
}
fail <- function(id, msg) respond(list(id = id, ok = FALSE, error = msg, trace = ""))

to_r <- function(x) {
  # JSON array of scalars -> R vector; nested/named stays list
  if (is.list(x)) {
    if (length(x) > 0 && is.null(names(x)) &&
        all(vapply(x, function(e) is.atomic(e) && length(e) == 1, TRUE)))
      return(unlist(x))
    return(lapply(x, to_r))
  }
  x
}

sym_import <- function(target) {
  # CRAN/library → ./<t>.R → ./rlib/<t>.R
  if (suppressWarnings(require(target, character.only = TRUE, quietly = TRUE)))
    return(list(package = target))
  for (f in c(paste0("./", target, ".R"), paste0("./rlib/", target, ".R"))) {
    if (file.exists(f)) { source(f); return(list(file = f)) }
  }
  stop(paste0("cannot import '", target, "' (no package, ./", target, ".R, ./rlib/)"))
}

sym_call <- function(target, args) {
  # "pkg::fn" or "fn" — R functions take positional args
  fn <- eval(parse(text = gsub("\\.", "::", target, fixed = FALSE)))
  if (!is.function(fn)) fn <- get(target)
  do.call(fn, args)
}

HANDLES <- new.env()
HID <- 0L

handle_for <- function(obj) {
  HID <<- HID + 1L
  assign(as.character(HID), obj, envir = HANDLES)
  list("__sym__" = "handle", runtime = "r", id = HID,
       type = paste(class(obj), collapse = ","))
}
deref_args <- function(args) {
  lapply(args, function(a) {
    if (is.list(a) && identical(a[["__sym__"]], "handle"))
      get(as.character(a$id), envir = HANDLES)
    else to_r(a)
  })
}
to_symbol <- function(v) {
  if (is.null(v) || is.atomic(v)) return(v)
  if (is.list(v) && is.null(attr(v, "class"))) return(lapply(v, to_symbol))
  handle_for(v)  # models, environments, S4 objects -> handles
}

env_sym <- new.env()

while (TRUE) {
  line <- readLines(con_in, n = 1)
  if (length(line) == 0) break
  line <- trimws(line)
  if (nchar(line) == 0) next
  msg <- tryCatch(fromJSON(line, simplifyVector = FALSE), error = function(e) NULL)
  if (is.null(msg)) next
  id <- msg$id
  result <- tryCatch({
    if (msg$op == "ping") ok(id, "pong")
    else if (msg$op == "shutdown") { ok(id, "bye"); quit(save = "no") }
    else if (msg$op == "import") ok(id, sym_import(msg$target))
    else if (msg$op == "call") {
      args <- if (is.null(msg$args)) list() else deref_args(msg$args)
      ok(id, to_symbol(sym_call(msg$target, args)))
    }
    else if (msg$op == "hcall") {
      obj <- get(as.character(msg$handle), envir = HANDLES)
      args <- if (is.null(msg$args)) list() else deref_args(msg$args)
      fn <- get(msg$method)
      ok(id, to_symbol(do.call(fn, c(list(obj), args))))
    }
    else if (msg$op == "free") {
      rm(list = as.character(msg$handle), envir = HANDLES); ok(id, TRUE)
    }
    else if (msg$op == "exec") {
      sym <- if (is.null(msg$env)) list() else lapply(msg$env, to_r)
      assign("sym", sym, envir = env_sym)
      value <- eval(parse(text = msg$code), envir = env_sym)
      exports <- lapply(get("sym", envir = env_sym), to_symbol)
      ok(id, to_symbol(value), exports)
    }
    else fail(id, paste0("unknown op '", msg$op, "'"))
  }, error = function(e) fail(id, conditionMessage(e)))
}

File: bridge/workers/worker.php
<?php
// SymBridge PHP worker — a PHP runtime conducted by Sym.
// JSON lines over stdin/stdout. Talks only to Sym.

error_reporting(E_ALL & ~E_DEPRECATED & ~E_NOTICE & ~E_WARNING);
$GLOBALS['__modules'] = [];
$GLOBALS['__handles'] = [];
$GLOBALS['__hid'] = 0;

function handle_for($obj) {
    $id = ++$GLOBALS['__hid'];
    $GLOBALS['__handles'][$id] = $obj;
    return ["__sym__" => "handle", "runtime" => "php", "id" => $id,
            "type" => get_class($obj)];
}
function deref($args) {
    $out = [];
    foreach ($args as $a) {
        if (is_array($a) && ($a["__sym__"] ?? "") === "handle") {
            $o = $GLOBALS['__handles'][$a["id"]] ?? null;
            if ($o === null) throw new Exception("stale handle #" . $a["id"]);
            $out[] = $o;
        } else $out[] = $a;
    }
    return $out;
}
function to_symbol($v) {
    if (is_object($v) && !($v instanceof stdClass)) return handle_for($v);
    if (is_array($v)) { $o = []; foreach ($v as $k => $x) $o[$k] = to_symbol($x); return $o; }
    return $v;
}

function respond($a) { fwrite(STDOUT, json_encode($a) . "\n"); fflush(STDOUT); }
function ok($id, $value = null, $exports = null) {
    $r = ["id" => $id, "ok" => true, "value" => $value];
    if ($exports !== null) $r["exports"] = $exports;
    respond($r);
}
function fail($id, $e) {
    respond(["id" => $id, "ok" => false,
             "error" => $e->getMessage(), "trace" => $e->getTraceAsString()]);
}

function sym_import($target) {
    // resolution: composer autoload → ./<t>.php → ./phplib/<t>.php → builtin ext
    foreach (["vendor/autoload.php"] as $auto)
        if (file_exists($auto)) require_once $auto;
    foreach (["./$target.php", "./phplib/$target.php"] as $f)
        if (file_exists($f)) { require_once $f; return ["file" => $f]; }
    if (class_exists($target) || function_exists($target) || extension_loaded($target))
        return ["builtin" => $target];
    throw new Exception("cannot import '$target' (no composer entry, ./$target.php, ./phplib/$target.php, class or extension)");
}

function sym_call($target, $args) {
    // "Class::method", "Class.method", plain function, or namespaced
    $target = str_replace(".", "::", $target);
    if (strpos($target, "::") !== false) {
        [$cls, $m] = explode("::", $target, 2);
        if (!class_exists($cls)) throw new Exception("class '$cls' not found");
        return call_user_func_array([$cls, $m], $args);
    }
    if (!is_callable($target)) throw new Exception("'$target' is not callable");
    return call_user_func_array($target, $args);
}

$stdin = fopen("php://stdin", "r");
while (($line = fgets($stdin)) !== false) {
    $line = trim($line);
    if ($line === "") continue;
    $msg = json_decode($line, true);
    if (!$msg) continue;
    $id = $msg["id"] ?? null;
    try {
        switch ($msg["op"] ?? "") {
            case "ping":     ok($id, "pong"); break;
            case "shutdown": ok($id, "bye"); exit(0);
            case "import":   ok($id, sym_import($msg["target"])); break;
            case "call":     ok($id, to_symbol(sym_call($msg["target"], deref($msg["args"] ?? [])))); break;
            case "new": {
                $cls = str_replace(".", "\\", $msg["target"]);
                if (!class_exists($cls)) throw new Exception("class '$cls' not found");
                $obj = new $cls(...deref($msg["args"] ?? []));
                ok($id, handle_for($obj)); break;
            }
            case "hcall": {
                $obj = $GLOBALS['__handles'][$msg["handle"]] ?? null;
                if ($obj === null) throw new Exception("stale handle #" . $msg["handle"]);
                $m = $msg["method"];
                ok($id, to_symbol($obj->$m(...deref($msg["args"] ?? [])))); break;
            }
            case "free": unset($GLOBALS['__handles'][$msg["handle"]]); ok($id, true); break;
            case "exec":
                $sym = $msg["env"] ?? [];
                $__code = $msg["code"];
                $value = (function () use (&$sym, $__code) {
                    return eval($__code);
                })();
                ok($id, $value, $sym);
                break;
            default: throw new Exception("unknown op");
        }
    } catch (Throwable $e) { fail($id, $e); }
}
